File: Problemset/maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.java

// @Title: 串联字符串的最大长度 (Maximum Length of a Concatenated String with Unique Characters)
// @Author: cocofe
// @Date: 2021-06-19 02:55:36
// @Runtime: 3 ms
// @Memory: 35.9 MB

class Solution {
    int ans = 0;
    public int maxLength(List<String> arr) {
        LinkedList<Integer> masks = new LinkedList();
        for(String s:arr) {
            int mask = 0;
            for(int i=0; i<s.length();i++){
                int offset = s.charAt(i) - 'a';
                if (((mask >> offset) & 1) == 1){
                    mask=0;
                    break;
                }
                mask |= 1 << offset;
            }
            if (mask > 0){
                masks.add(mask);
            }
        }
        // System.out.println(masks);
        backtrace(0, 0, masks);
        return ans;
    }
    public void backtrace(int idx, int mask, LinkedList<Integer> masks) {
        if(idx == masks.size()){
            ans = Math.max(ans, Integer.bitCount(mask));
            return;
        }
        // System.out.println("" + (mask ^ masks.get(idx)) + "|" + mask + "|" + masks.get(idx));

        if((mask & masks.get(idx)) == 0) {
            backtrace(idx+1, mask | masks.get(idx), masks);
        }
        backtrace(idx+1, mask, masks);

    }
}

File: Problemset/maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.py

# @Title: 串联字符串的最大长度 (Maximum Length of a Concatenated String with Unique Characters)
# @Author: cocofe
# @Date: 2021-06-19 02:28:48
# @Runtime: 96 ms
# @Memory: 15.2 MB

class Solution:
    def maxLength(self, arr: List[str]) -> int:
        # 要清楚是数组的子序列拼接成字符串, 不是数组元素的子序列互相拼接
        masks = []
        for s in arr:
            mask = 0
            for ch in s:
                if mask >> (ord(ch) - ord('a')) & 1 == 1:
                    mask = 0
                    break
                mask |= 1 << (ord(ch) - ord('a'))
            if mask:
                masks.append(mask)
        ans = 0
        def backtrace(idx, mask):
            if idx == len(masks):
                nonlocal ans
                ans = max(ans, bin(mask).count('1'))
                return
            if mask == 0 or (masks[idx] ^ mask) & mask == mask:
                backtrace(idx+1, mask | masks[idx])
            backtrace(idx+1, mask)
        backtrace(0, 0)
        return ans
        

            
